File: src/cleansia_ios/CleansiaPartner/Sources/Features/Orders/WorkContract.swift
import CleansiaCore
import CleansiaPartnerApi
import Foundation

/// A contract for work rendered for one job: the text row the take echoes, the facts the acceptance
/// binds, and — on a read of an accepted contract — the acceptance behind it. A preview carries no
/// `acceptance`; a read carries the facts as they were stored, never the live order.
struct WorkContract: Equatable {
    let legalDocumentTextId: String
    let version: String
    let language: String?
    let title: String?
    let contentHtml: String
    let facts: WorkContractJobFacts
    let acceptance: WorkContractAcceptanceFacts?
}

struct WorkContractJobFacts: Equatable {
    let orderNumber: String?
    let cleaningDateTimeUtc: Date
    let estimatedMinutes: Int
    let totalPrice: Double
    let currencyCode: String?
    let locationApproximate: String?
    let rooms: Int
    let bathrooms: Int
    let services: [String]
    let packages: [String]
    let extraSlugs: [String]
}

struct WorkContractAcceptanceFacts: Equatable {
    let acceptedOn: Date
    let documentVersion: String
    let acceptedLanguage: String?
}

extension WorkContract {
    /// **Refuse.** The text id is what the take echoes, the HTML is what is accepted, and the price,
    /// window and scope are what the acceptance binds — a null in any of them is a broken wire, and a
    /// sheet that showed "0 Kč" over an empty text would record an acceptance of nothing. The labels
    /// (number, currency code, location) stay nullable and render as absent.
    init(_ dto: WorkContractDto) throws {
        legalDocumentTextId = try dto.legalDocumentTextId.requireNonBlank("legalDocumentTextId")
        version = try dto.version.requireNonBlank("version")
        language = dto.language
        title = dto.title
        contentHtml = try dto.contentHtml.require("contentHtml")
        facts = try WorkContractJobFacts(dto.facts.require("facts"))
        acceptance = try dto.acceptance.map(WorkContractAcceptanceFacts.init)
    }
}

extension WorkContractJobFacts {
    init(_ dto: WorkContractFacts) throws {
        orderNumber = dto.orderNumber
        cleaningDateTimeUtc = try dto.cleaningDateTimeUtc.require("facts.cleaningDateTimeUtc")
        estimatedMinutes = try dto.estimatedMinutes.require("facts.estimatedMinutes")
        totalPrice = try dto.totalPrice.require("facts.totalPrice")
        currencyCode = dto.currencyCode
        locationApproximate = dto.locationApproximate
        rooms = try dto.rooms.require("facts.rooms")
        bathrooms = try dto.bathrooms.require("facts.bathrooms")
        services = Self.names(dto.services)
        packages = Self.names(dto.packages)
        extraSlugs = dto.extraSlugs ?? []
    }

    private static func names(_ lines: [WorkContractFactsLine]?) -> [String] {
        (lines ?? []).compactMap { line in
            guard let name = line.name, !name.isBlank else { return nil }
            return name
        }
    }
}

extension WorkContractAcceptanceFacts {
    init(_ dto: WorkContractAcceptanceDetails) throws {
        acceptedOn = try dto.acceptedOn.require("acceptance.acceptedOn")
        documentVersion = try dto.documentVersion.requireNonBlank("acceptance.documentVersion")
        acceptedLanguage = dto.acceptedLanguage
    }
}

File: src/cleansia_android/partner-app/src/main/java/cz/cleansia/partner/data/orders/WorkContract.kt
package cz.cleansia.partner.data.orders

import cz.cleansia.core.network.required
import cz.cleansia.partner.api.model.WorkContractAcceptanceDetails
import cz.cleansia.partner.api.model.WorkContractDto
import cz.cleansia.partner.api.model.WorkContractFacts
import cz.cleansia.partner.api.model.WorkContractFactsLine

/**
 * A contract for work rendered for one job: the text row the take echoes, the facts the acceptance
 * binds, and — on a read of an accepted contract — the acceptance behind it. A preview carries no
 * [acceptance]; a read carries the facts as they were stored, never the live order.
 */
data class WorkContract(
    val legalDocumentTextId: String,
    val version: String,
    val language: String?,
    val title: String?,
    val contentHtml: String,
    val facts: WorkContractJobFacts,
    val acceptance: WorkContractAcceptanceFacts?,
)

data class WorkContractJobFacts(
    val orderNumber: String?,
    val cleaningDateTimeUtc: String,
    val estimatedMinutes: Int,
    val totalPrice: Double,
    val currencyCode: String?,
    val locationApproximate: String?,
    val rooms: Int,
    val bathrooms: Int,
    val services: List<String>,
    val packages: List<String>,
    val extraSlugs: List<String>,
)

data class WorkContractAcceptanceFacts(
    val acceptedOn: String,
    val documentVersion: String,
    val acceptedLanguage: String?,
)

/**
 * Refuses rather than defaults: the text id is what the take echoes, the HTML is what is accepted,
 * and the price, window and scope are what the acceptance binds — a null in any of them is a broken
 * wire, and a sheet that showed "0 Kč" over an empty text would record an acceptance of nothing.
 * The labels (number, currency code, location) stay nullable and render as absent.
 */
internal fun WorkContractDto.toDomain(): WorkContract = WorkContract(
    legalDocumentTextId = legalDocumentTextId.required("legalDocumentTextId"),
    version = version.required("version"),
    language = language,
    title = title,
    contentHtml = contentHtml.required("contentHtml"),
    facts = facts.required("facts").toDomain(),
    acceptance = acceptance?.toDomain(),
)

private fun WorkContractFacts.toDomain(): WorkContractJobFacts = WorkContractJobFacts(
    orderNumber = orderNumber,
    cleaningDateTimeUtc = cleaningDateTimeUtc.required("facts.cleaningDateTimeUtc"),
    estimatedMinutes = estimatedMinutes.required("facts.estimatedMinutes"),
    totalPrice = totalPrice.required("facts.totalPrice"),
    currencyCode = currencyCode,
    locationApproximate = locationApproximate,
    rooms = rooms.required("facts.rooms"),
    bathrooms = bathrooms.required("facts.bathrooms"),
    services = services.orEmpty().names(),
    packages = packages.orEmpty().names(),
    extraSlugs = extraSlugs.orEmpty(),
)

private fun List<WorkContractFactsLine>.names(): List<String> = mapNotNull { it.name?.takeIf(String::isNotBlank) }

private fun WorkContractAcceptanceDetails.toDomain(): WorkContractAcceptanceFacts = WorkContractAcceptanceFacts(
    acceptedOn = acceptedOn.required("acceptance.acceptedOn"),
    documentVersion = documentVersion.required("acceptance.documentVersion"),
    acceptedLanguage = acceptedLanguage,
)
